### PR TITLE
Add energy ratio region

### PR DESCRIPTION
**Pull request by paulf81**
**Complete this sentence**
THIS PULL REQUEST __ IS __ READY TO MERGE

**Feature or improvement description**
- Adds additional energy ratio functions to the module
- Adds an example on usage of energy ratio using FLORIS
- Adds documentation

### CODE REVIEW

File: examples/example_0030_demo_energy_ratio.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import matplotlib.pyplot as plt
+import floris.tools as wfct
+import floris.tools.visualization as vis
+from floris.tools.energy_ratio import plot_energy_ratio, plot_energy_ratio_ws 
+import floris.tools.cut_plane as cp
+import numpy as np
+from mpl_toolkits.axes_grid1 import make_axes_locatable, axes_size
+
+# Parameters
+n_sim = 1 # number of simulations to use
```
Comment (ejsimley) [inline examples/example_0030_demo_energy_ratio.py]: Suggest changing default n_sim to ~1000 to get more meaningful plots

File: floris/tools/energy_ratio.py [hunk 2]
```diff
@@ -48,6 +48,283 @@ def _get_confidence_bounds(confidence):
...
+            -   **counts_pchange** (*float*): Number of points in 
+                pchange (min(baseline,controlled)).
+    """
+
+    # Combine the ws/wd into tuples
+    base_tuples = np.array(list(zip(ws_base,wd_base)))
+    
+    # base_set = set(base_tuples)
+    con_tuples = np.array(list(zip(ws_con,wd_con)))
+    con_tuples[0,:] = [4,333]
```
Comment (ejsimley) [inline floris/tools/energy_ratio.py]: Is this left over from debugging?

File: floris/tools/energy_ratio.py [hunk 3]
```diff
@@ -48,6 +48,283 @@ def _get_confidence_bounds(confidence):
     return [50 + 0.5 * confidence, 50 - 0.5 * confidence]
 
 
+def energy_ratio_region(ref_pow_base, test_pow_base, ws_base,wd_base,
+                 ref_pow_con, test_pow_con, ws_con, wd_con, wd_step=1):
```
Comment (ejsimley) [inline floris/tools/energy_ratio.py]: wd_step doesn't seem to be used

File: floris/tools/energy_ratio.py [hunk 4]
```diff
@@ -48,6 +48,283 @@ def _get_confidence_bounds(confidence):
...
+                                    wind_direction_array_baseline,
+                                    reference_power_controlled,
+                                    test_power_controlled,
+                                    wind_speed_array_controlled,
+                                    wind_direction_array_controlled,
+                                    confidence=95,
+                                    n_boostrap=None
+                                    ):
+    """
+    Calculate a balanced energy ratio over a region
```
Comment (ejsimley) [inline floris/tools/energy_ratio.py]: doc string should be updated. Example: get rid of wind_direction_bins description

File: floris/tools/energy_ratio.py [hunk 5]
```diff
@@ -478,6 +753,541 @@ def plot_energy_ratio(reference_power_baseline,
...
+    energy ratio calculation for a particular wind direction bin.  Note 
+    the reference turbine should not be the turbine implementing 
+    control, but should be an unaffected nearby turbine, or a synthetic 
+    power estimate from a measurement.
+
+    Args:
+        ref_pow_base (np.array): Array of baseline reference turbine 
+            power.
+        test_pow_base (np.array): Array of baseline test turbine power.
+        wd_base (np.array): Array of wind speeds for basline.
```
Comment (ejsimley) [inline floris/tools/energy_ratio.py]: Update documentation, example: Array of wind "directions" for baseline

File: floris/tools/energy_ratio.py [hunk 6]
```diff
@@ -478,6 +753,541 @@ def plot_energy_ratio(reference_power_baseline,
...
+
+#######WIND SPEED VERSIONS##################
+
+def energy_ratio_ws(ref_pow_base, test_pow_base, wd_base,
+                 ref_pow_con, test_pow_con, wd_con, use_absolutes=False,use_mean=False):
+    """
+    Compute the balanced energy ratio
+
+    This function is typically called to compute a single balanced 
+    energy ratio calculation for a particular wind direction bin.  Note
```
Comment (ejsimley) [inline floris/tools/energy_ratio.py]: for a particular wind "speed" bin?

File: floris/tools/energy_ratio.py [hunk 7]
```diff
@@ -478,6 +753,541 @@ def plot_energy_ratio(reference_power_baseline,
...
+                                    wind_direction_array_controlled,
+                                    wind_speed_bins,
+                                    confidence=95,
+                                    n_boostrap=None,
+                                    wind_speed_bin_p_overlap=None,
+                                    use_absolutes=False,
+                                    use_mean=False
+                                    ):
+    """
+    Calculate a balanced energy ratio for each wind direction bin.
```
Comment (ejsimley) [inline floris/tools/energy_ratio.py]: In general, doc strings should be updated for the new wind speed functions

**Comment by paulf81** [comment]
Nice catches @ejsimley , this next batch of commits I hope fixes the issues...

**Pull request merged**